File: 22.10_PS/1022_거의_소수.java
// boj. 1456

import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        long a = Long.parseLong(st.nextToken());
        long b = Long.parseLong(st.nextToken());
        int c = (int)Math.sqrt(b);
        long answer = 0;

        long[] arr = new long[c + 1];
        arr[0] = 1;
        arr[1] = 1;

        for (int i = 2; i < Math.sqrt(c); i++) {
            if (arr[i] == 0) {
                int num = 2;
                while (i * num <= c) {
                    arr[i * num] = 1;
                    num++;
                }
            }
        }
수
        for (int j = 2; j <= c; j++) {
            double powNum = 2;
            if (arr[j] == 0) {
                while (Math.pow(j, powNum) <= b) {
                    if (Math.pow(j, powNum) >= a)
                        answer++;
                    powNum++;
                }
            }
        }
        System.out.println(answer);
    }
}

File: 22.10_PS/1022_거의_소수.py
# boj. 1456

import sys
import math

input = sys.stdin.readline
a, b = map(int, input().split())
c = int(math.sqrt(b))
count = 0
arr = [True] * (c + 1)
arr[0], arr[1] = False, False

# 에라토스테네스의 체(소수인 경우 True)
for i in range(2, int(math.sqrt(c)) + 1):
    num = 2
    if arr[i]:
        while i * num <= c:
            arr[i * num] = False
            num += 1

# 소수 배열에서 거의 소수 탐색
for i in range(2, c + 1):
    if arr[i]:
        num = 2
        while math.pow(i, num) <= b:
            t = math.pow(i, num)
            if t >= a: # 제곱수의 범위가 a이하인 경우는 num 가산만 수행
                count += 1
            num += 1

print(count)
